### PR TITLE
LOTR: Lord of the Retrievers. A retriever that merge several retrievers together applying document_formatters to them.

### PR DESCRIPTION
**Pull request by GMartin-dev**
"One Retriever to merge them all, One Retriever to expose them, One Retriever to bring them all and in and process them with Document formatters."

Hi @dev2049! Here bothering people again!

I'm using this simple idea to deal with merging the output of several retrievers into one.
I'm aware of DocumentCompressorPipeline and ContextualCompressionRetriever but I don't think they allow us to do something like this. Also I was getting in trouble to get the pipeline working too. Please correct me if i'm wrong.

This allow to do some sort of "retrieval" preprocessing and then using the retrieval with the curated results anywhere you could use a retriever.
My use case is to generate diff indexes with diff embeddings and sources for a more colorful results then filtering them with one or many document formatters.

I saw some people looking for something like this, here:
https://github.com/hwchase17/langchain/issues/3991
and something similar here:
https://github.com/hwchase17/langchain/issues/5555

This is just a proposal I know I'm missing tests , etc. If you think this is a worth it idea I can work on tests and anything you want to change.
Let me know!

### CODE REVIEW

File: langchain/retrievers/merger_retriever.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        Merge the results of the retrievers.
+
+        Args:
+            query: The query to search for.
+
+        Returns:
+            A list of merged documents.
+        """
+
+        # Get the results of all retrievers.
```
Comment (cancan101) [inline langchain/retrievers/merger_retriever.py:81]: not sure how much better this is, but from https://docs.python.org/3/library/itertools.html#itertools-recipes:
```python
def roundrobin(*iterables):
    "roundrobin('ABC', 'D', 'EF') --> A D E B F C"
    # Recipe credited to George Sakkis
    num_active = len(iterables)
    nexts = cycle(iter(it).__next__ for it in iterables)
    while num_active:
        try:
            for next in nexts:
                yield next()
        except StopIteration:
            # Remove the iterator we just exhausted from the cycle.
            num_active -= 1
            nexts = cycle(islice(nexts, num_active))
```

File: langchain/retrievers/merger_retriever.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+        # Merge the results of the retrievers.
+        merged_documents = await self.amerge_documents(query)
+
+        # Refine the results using the transformers.
+        refined_documents = await self.arefine_documents(merged_documents)
+
+        return refined_documents
+
+    def merge_documents(self, query: str) -> List[str]:
```
Comment (cancan101) [inline langchain/retrievers/merger_retriever.py:70]: could make this return `Iterable[str]`

File: langchain/retrievers/merger_retriever.py (hunk 1 above)
Comment (GDrupal) [reply langchain/retrievers/merger_retriever.py:81]: That's fancy logic right there! I never thought this was a roundrobin use case but  if kinda fits. 
Downside: It's a little harder to read.
I was thinking that we could expose a couple of diff merging strategies with a configurable parameter.

**Review by hwchase17 — COMMENTED** [review]
this is pretty cool (and I love the name)

how does this compare to `ContextualCompressionRetriever`? `base_compressor` can already be a list of transformers (DocumentCompressorPipeline). so then the main new thing here seems to be accepting a list of retrievers instead of a single retriever? if thats case, maybe we could just add a new class which is a retriever over multiple retrievers? and then this functionality can be achieved by combining that in a ContextualCompressionRetriever with a DocumentCompressorPipeline

a notebook would also help a lot

**Comment by GDrupal** [comment]
@hwchase17 
Yeap you are right you can do the document formatter part with the compressor pipeline. 
I was having issues getting it working but I just tried it with the merger retriever and works.
I will remove the document formatter part and add some tests.
Here an example on how would look:
```Python

# Define 3 diff collections
client_settings = chromadb.config.Settings(
    chroma_db_impl="duckdb+parquet",
    persist_directory=DB_DIR,
    anonymized_telemetry=False,
)
vxsumall = Chroma(
    collection_name="project_store_all",
    persist_directory=DB_DIR,
    client_settings=client_settings,
    embedding_function=emall,
)
vxsummulti = Chroma(
    collection_name="project_store_multi",
    persist_directory=DB_DIR,
    client_settings=client_settings,
    embedding_function=emmulti,
)
vnosumall = Chroma(
    collection_name="project_store_nosum",
    persist_directory=DB_DIR,
    client_settings=client_settings,
    embedding_function=emall,
)

# Define 3 diff retrievers
ret0 = vnosumall.as_retriever(
    search_type="similarity", search_kwargs={"k": 5, "include_metadata": True}
)
ret1 = vxsumall.as_retriever(
    search_type="similarity", search_kwargs={"k": 10, "include_metadata": True}
)
ret2 = vxsummulti.as_retriever(
    search_type="mmr", search_kwargs={"k": 10, "include_metadata": True}
)
em_filter2 = EmbeddingsRedundantFilter(embeddings=emall)

# We just pass a list of retrievers.
merger = MergerRetriever(retrievers=[ret0, ret1, ret2])

# And if we want to clean the redundant documents "overlap" between the 3 retrievers.
pipeline_comp = DocumentCompressorPipeline(transformers=[em_filter2])
comp_ret1 = ContextualCompressionRetriever(
    base_compressor=pipeline_comp, base_retriever=merger
)

```

**Comment by hwchase17** [comment]
@GDrupal perfect. a notebook would also help. excited for this!

**Comment by simonfromla** [comment]
Gave it a try and it works well! Looking forward to the merge :)

**Comment by hwchase17** [comment]
gonna save this one for a special saturday release and give it a full release thread! @GDrupal is there a good twitter handle to give you an appropriate shout out on?

**Comment by GDrupal** [comment]
@hwchase17 Awesome! I do not have a formal one as a developer but you can use @musicaoriginal2 (I write some orchestral music on the side xD)
Are you planning to do Linkedin posts along with Twitter? 
My only "formal" dev profile out there it's at: https://www.linkedin.com/in/gmartindev/

**Comment by ghoshk24** [comment]
@GDrupal Just tried it out. Works perfectly. Really helpful! :)

**Comment by homanp** [comment]
@GDrupal is this possible to use with ConversationalRetrievalChain?

**Comment by GDrupal** [comment]
> @GDrupal is this possible to use with ConversationalRetrievalChain?

Yeap it works transparently as any other retriever

**Comment by homanp** [comment]
> > @GDrupal is this possible to use with ConversationalRetrievalChain?
> 
> Yeap it works transparently as any other retriever

I'm doing this currently but can't seem to get anything into the context of the prompt:

```
document_retrievers = MergerRetriever(retrievers)
question_generator = LLMChain(
    llm=OpenAI(temperature=0), prompt=CONDENSE_QUESTION_PROMPT
)
doc_chain = load_qa_chain(
    llm,
    chain_type="stuff",
    prompt=QA_PROMPT,
    verbose=True,
)
agent = ConversationalRetrievalChain(
    retriever=document_retrievers,
    combine_docs_chain=doc_chain,
    question_generator=question_generator,
    memory=memory,
    get_chat_history=lambda h: h,
    output_key="output",
)
```

Do you see anything wrong in this?

**Comment by GDrupal** [comment]
Did you confirmed that the pinecone retrievers generate results? 
are CONDENSE_QUESTION_PROMPT and QA_PROMPT custom?

**Comment by homanp** [comment]
@GDrupal 

The retrievers list is as following and generate results.:
```
[VectorStoreRetriever(vectorstore=<langchain.vectorstores.pinecone.Pinecone object at 0x1160aa430>, search_type='similarity', search_kwargs={}), VectorStoreRetriever(vectorstore=<langchain.vectorstores.pinecone.Pinecone object at 0x115dc5640>, search_type='similarity', search_kwargs={})]
```

 The `QA_PROMPT` is unchanged, the same for `CONDENSE_QUESTION_PROMPT `:
 
 ```
 qa_prompt_template = """Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.

{context}

Question: {question}
Helpful Answer:"""

QA_PROMPT = PromptTemplate(
    template=qa_prompt_template, input_variables=["context", "question"]
)
```

**Comment by homanp** [comment]
@GDrupal nevermind, issue on my side! Thanks for your support <3

**Pull request merged**